features/nvstore/source/nvstore_shared_lock.h

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2018 ARM Limited. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+
+#ifndef __NVSTORE_SHARED_LOCK_H
+#define __NVSTORE_SHARED_LOCK_H
+
+#include <stdint.h>
+#include "rtos/Mutex.h"
+
+enum {
+    NVSTORE_OS_OK          =  0,
+    NVSTORE_OS_RTOS_ERR    = -1,
+    NVSTORE_OS_INV_ARG_ERR = -2,
+    NVSTORE_OS_NO_MEM_ERR  = -3,
+};
+
+/**
+ * SharedLock, implements a shared/exclusive lock (AKA RW lock).
+ * This class doesn't implement this kind of lock in the classic manner (having two Mutexes and a counter).
+ * Instead, it uses one Mutex and one atomic counter, favouring the shared usage of the exclusive one.
+ */
+
+class SharedLock {
```

Comment:
@kjbracey-arm Can you review this `SharedLock` class ?

Reply:
Thanks for the thorough review. This was indeed implemented privately for NVStore, with the intention of favoring the shared usage (writers) over the exclusive one (GC), which is why I didn't use the classic two mutexes implementation. Guess that if there was such an mbed-os class, I'd use it instead. Anyway I'll add an NVStore namespace there to keep it bounded to NVStore.

Reply:
Fine. But do check my top-level PR comments that I failed to attach to the diff - I think there may be a couple of flaws in how it's used by nvstore.